# Terragrunt multiple accounts with assume role getting connection refused from aws

**wiseelf** commented *Aug 30, 2022*

Terrafrom version: 1.2.8
Terragrunt version: 0.38.9
AWS provider version: 4.28.0

My config:
```
iam_role = local.iam_role

# Generate an AWS provider block
generate "provider" {
  path = "provider.tf"
  if_exists = "overwrite_terragrunt"
  contents = <<EOF
provider "aws" {
  region  = "${local.aws_region}"

  # Only these AWS Account IDs may be operated on by this template
  allowed_account_ids = ["${local.account_id}"]
}
EOF
}

# Configure Terragrunt to automatically store tfstate files in an S3 bucket
remote_state {
  backend = "s3"
  config = {
    encrypt        = true
    bucket         = "terraform-st-acc-${local.account_id}-${local.aws_region}"
    key            = "${path_relative_to_include()}/terraform.tfstate"
    region         = local.aws_region
    dynamodb_table = "terraform-locks"
  }
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }
}
```

Having multiple accounts and switching between them with assume-role.
During `terragrunt run-all apply --terragrunt-non-interactive --terragrunt-parallelism 5` getting connection refused errors from sts, dynamodb, s3, etc. Example:
```
ERRO[0040] Create S3 bucket with retry iris-terraform-st-acc-1234567890-us-east-1 returned an error: RequestError: send request failed
caused by: Put "https://terraform-st-acc-1234567890-us-east-1.s3.amazonaws.com/": dial tcp 52.217.133.121:443: connect: connection refused. Sleeping for 10s and will try again.
ERRO[0044] Create S3 bucket with retry terraform-st-acc-1234567890-us-east-1 returned an error: RequestError: send request failed
caused by: Put "https://terraform-st-acc-1234567890-us-east-1.s3.amazonaws.com/?versioning=": dial tcp 52.217.133.121:443: connect: connection refused. Sleeping for 10s and will try again. 

Initializing the backend...
╷
│ Error: error configuring S3 Backend: error validating provider credentials: error calling sts:GetCallerIdentity: RequestError: send request failed
│ caused by: Post "https://sts.amazonaws.com/": dial tcp 209.54.180.124:443: connect: connection refused
│
│

│ Error: Error releasing the state lock
│
│ Error message: RequestError: send request failed
│ caused by: Post "https://dynamodb.us-east-1.amazonaws.com/": dial tcp
│ 3.218.182.10:443: connect: connection refused
│ Lock Info:
```

Looks like terragrunt is doing sts:assume for each operation such as: to read info from backend, to put a lock, to remove a lock. So AWS basically starts throttling my requests and i'm getting connection refused errors. 
Am I missing something? Is there a way to avoid these errors and keep using assume-roles? 
<br />
***


**wiseelf** commented *Nov 1, 2022*

Recently I reworked my terragrunt code and instead of using separate folders for each role,policy, etc I combined them into one module, looks like this:
Before:
```
account/region/iam/policies/policy_name1/terragrunt.hcl
account/region/iam/policies/policy_name2/terragrunt.hcl
account/region/iam/policies/policy_name3/terragrunt.hcl
account/region/iam/roles/role_name1/terragrunt.hcl
account/region/iam/roles/role_name2/terragrunt.hcl
account/region/iam/users/user_name1/terragrunt.hcl
account/region/iam/users/user_name2/terragrunt.hcl
```

Now:
```
account/region/iam/policies/terragrunt.hcl
account/region/iam/roles/terragrunt.hcl
account/region/iam/users/terragrunt.hcl
```
Even with this changes since my infra is still big I anyway continue getting this error:
```
finished with an error: RequestError: send request failed
caused by: Post "https://sts.amazonaws.com/": dial tcp 54.239.29.25:443: connect: connection refused
```

Please introduce some sort of STS session caching. Or please tell me how to properly use sts assume role to not hit AWS rate limits. This guide https://terragrunt.gruntwork.io/docs/features/work-with-multiple-aws-accounts/ works only if you have very simple setup.


Current versions:
terraform_version_constraint = "= 1.3.3"
terragrunt_version_constraint = "= 0.39.2"
aws_provider_version        = "4.36.0"
***

**lorengordon** commented *Nov 1, 2022*

As a possible workaround, you might try adding that to the list of retryable errors?

```
retryable_errors = [
  # Terragrunt default retryable errors
  "(?s).*Failed to load state.*tcp.*timeout.*",
  "(?s).*Failed to load backend.*TLS handshake timeout.*",
  "(?s).*Creating metric alarm failed.*request to update this alarm is in progress.*",
  "(?s).*Error installing provider.*TLS handshake timeout.*",
  "(?s).*Error configuring the backend.*TLS handshake timeout.*",
  "(?s).*Error installing provider.*tcp.*timeout.*",
  "(?s).*Error installing provider.*tcp.*connection reset by peer.*",
  "NoSuchBucket: The specified bucket does not exist",
  "(?s).*Error creating SSM parameter: TooManyUpdates:.*",
  "(?s).*app.terraform.io.*: 429 Too Many Requests.*",
  "(?s).*ssh_exchange_identification.*Connection closed by remote host.*",
  "(?s).*Client\\.Timeout exceeded while awaiting headers.*",
  "(?s).*Could not download module.*The requested URL returned error: 429.*",
  "(?s).*Error: NoCredentialProviders: no valid providers in chain.*",
  # Custom retryable errors
  "(?s).*dial tcp.*connect: connection refused.*",
]
```
***

**wiseelf** commented *Nov 2, 2022*

@lorengordon AWS applies blocking for 5-10 minutes. So it will not work, unfortunately. With a such workaround there is a chance to get even longer ban :)
***

